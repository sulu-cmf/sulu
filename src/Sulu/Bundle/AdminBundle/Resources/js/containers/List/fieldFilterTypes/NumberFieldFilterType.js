// @flow
import React from 'react';
import {computed} from 'mobx';
import Input from '../../../components/Input';
import SingleSelect from '../../../components/SingleSelect';
import AbstractFieldFilterType from './AbstractFieldFilterType';
import numberFieldFilterTypeStyles from './numberFieldFilterType.scss';

const operatorMapping = {
    lt: '<',
    eq: '=',
    gt: '>',
};

function getOperatorFromValue(value: ?{[string]: ?number}) {
    const valueKeys = value ? Object.keys(value) : {};

    if (valueKeys.length > 1) {
        throw new Error('The "NumberFilterFieldType" only accepts an array with exactly one key!');
    }

    return valueKeys[0];
}

function getNumberFromValue(value: ?{[string]: ?number}) {
    if (!value) {
        return undefined;
    }

    return value[Object.keys(value)[0]];
}

class NumberFieldFilterType extends AbstractFieldFilterType<?{[string]: ?number}> {
    constructor(
        onChange: (value: ?{[string]: ?number}) => void,
        parameters: ?{[string]: mixed},
        value: ?{[string]: ?number}
    ) {
        super(onChange, parameters, value);

        if (value === undefined) {
            onChange({eq: undefined});
        }
    }

    @computed get operator() {
        return getOperatorFromValue(this.value);
    }

    @computed get number() {
        return getNumberFromValue(this.value);
    }

    handleOperatorChange = (operatorValue: ?string) => {
        if (!operatorValue) {
            throw new Error('The operator cannot be changed to undefined! This should not happen and is likely a bug.');
        }

        const {onChange} = this;
        onChange({[operatorValue]: this.number});
    };

    handleInputChange = (inputValue: ?string) => {
        const {onChange} = this;
        onChange({[this.operator]: inputValue});
    };

    getFormNode() {
        return (
            <div className={numberFieldFilterTypeStyles.numberFieldFilterType}>
                <SingleSelect onChange={this.handleOperatorChange} value={this.operator}>
                    <SingleSelect.Option value="lt">{operatorMapping.lt}</SingleSelect.Option>
                    <SingleSelect.Option value="eq">{operatorMapping.eq}</SingleSelect.Option>
                    <SingleSelect.Option value="gt">{operatorMapping.gt}</SingleSelect.Option>
                </SingleSelect>
                <Input
                    onChange={this.handleInputChange}
                    type="number"
                    value={this.number}
                />
            </div>
        );
    }

    getValueNode(value: ?{[string]: ?number}) {
        return Promise.resolve(
            (operatorMapping[getOperatorFromValue(value)] || '') + ' ' + (getNumberFromValue(value) || '')
        );
    }
}

export default NumberFieldFilterType;
