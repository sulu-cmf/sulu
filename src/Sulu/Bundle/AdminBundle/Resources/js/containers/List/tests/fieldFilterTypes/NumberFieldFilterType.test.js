// @flow
import {mount, render} from 'enzyme';
import NumberFieldFilterType from '../../fieldFilterTypes/NumberFieldFilterType';

jest.mock('../../../../utils/Translator', () => ({
    translate: jest.fn((key) => key),
}));

test.each([
    [undefined],
    [{eq: 6}],
    [{lt: 2}],
    [{gt: 7}],
])('Render with value of "%s"', (value) => {
    const numberFieldFilterType = new NumberFieldFilterType(jest.fn(), {}, value);
    expect(render(numberFieldFilterType.getFormNode())).toMatchSnapshot();
});

test('Call onChange handler with a default operator when undefined is passed', () => {
    const changeSpy = jest.fn();
    new NumberFieldFilterType(changeSpy, {}, undefined);

    expect(changeSpy).toBeCalledWith({eq: undefined});
});

test.each([
    [{eq: 8}, 'lt', {lt: 8}],
    [{lt: 9}, 'gt', {gt: 9}],
    [{gt: 3}, 'eq', {eq: 3}],
])('Call onChange handler when the value was "%s" and the "%s" operator is chosen', (value, operator, newValue) => {
    const changeSpy = jest.fn();
    const numberFieldFilterType = new NumberFieldFilterType(changeSpy, {}, value);
    const numberFieldFilterTypeForm = mount(numberFieldFilterType.getFormNode());

    numberFieldFilterTypeForm.find('SingleSelect').prop('onChange')(operator);

    expect(changeSpy).toBeCalledWith(newValue);
});

test.each([
    [{eq: 8}, 3, {eq: 3}],
    [{lt: 9}, 4, {lt: 4}],
    [{gt: 3}, 9, {gt: 9}],
])('Call onChange handler when the value was "%s" and the number was changed to "%s"', (value, number, newValue) => {
    const changeSpy = jest.fn();
    const numberFieldFilterType = new NumberFieldFilterType(changeSpy, {}, value);
    const numberFieldFilterTypeForm = mount(numberFieldFilterType.getFormNode());

    numberFieldFilterTypeForm.find('Input').prop('onChange')(number);

    expect(changeSpy).toBeCalledWith(newValue);
});

test.each([
    [{eq: 8}, '= 8'],
    [{lt: 3}, '< 3'],
    [{gt: 9}, '> 9'],
    [undefined, ' '],
])('Return value node with value "%s"', (value, expectedValueNode) => {
    const numberFieldFilterType = new NumberFieldFilterType(jest.fn(), {}, undefined);

    const valueNodePromise = numberFieldFilterType.getValueNode(value);

    if (!valueNodePromise) {
        throw new Error('The getValueNode function must return a promise!');
    }

    return valueNodePromise.then((valueNode) => {
        expect(valueNode).toEqual(expectedValueNode);
    });
});
