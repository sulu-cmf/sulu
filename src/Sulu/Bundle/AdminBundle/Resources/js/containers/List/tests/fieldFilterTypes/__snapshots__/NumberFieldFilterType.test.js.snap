// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render with value of "{ eq: 6 }" 1`] = `
<div
  class="numberFieldFilterType"
>
  <div
    class="select"
    role="none"
  >
    <button
      class="displayValue default"
      type="button"
    >
      <div
        aria-label="="
        class="croppedText"
        title="="
      >
        <div
          aria-hidden="true"
          class="front"
        >
          =
        </div>
        <div
          aria-hidden="true"
          class="back"
        >
          <span />
        </div>
        <div
          class="whole"
        >
          =
        </div>
      </div>
      <span
        aria-label="su-angle-down"
        class="su-angle-down toggle"
      />
    </button>
  </div>
  <div
    class="input default left"
  >
    <input
      type="number"
      value="6"
    />
  </div>
</div>
`;

exports[`Render with value of "{ gt: 7 }" 1`] = `
<div
  class="numberFieldFilterType"
>
  <div
    class="select"
    role="none"
  >
    <button
      class="displayValue default"
      type="button"
    >
      <div
        aria-label=">"
        class="croppedText"
        title=">"
      >
        <div
          aria-hidden="true"
          class="front"
        >
          &gt;
        </div>
        <div
          aria-hidden="true"
          class="back"
        >
          <span />
        </div>
        <div
          class="whole"
        >
          &gt;
        </div>
      </div>
      <span
        aria-label="su-angle-down"
        class="su-angle-down toggle"
      />
    </button>
  </div>
  <div
    class="input default left"
  >
    <input
      type="number"
      value="7"
    />
  </div>
</div>
`;

exports[`Render with value of "{ lt: 2 }" 1`] = `
<div
  class="numberFieldFilterType"
>
  <div
    class="select"
    role="none"
  >
    <button
      class="displayValue default"
      type="button"
    >
      <div
        aria-label="<"
        class="croppedText"
        title="<"
      >
        <div
          aria-hidden="true"
          class="front"
        >
          &lt;
        </div>
        <div
          aria-hidden="true"
          class="back"
        >
          <span />
        </div>
        <div
          class="whole"
        >
          &lt;
        </div>
      </div>
      <span
        aria-label="su-angle-down"
        class="su-angle-down toggle"
      />
    </button>
  </div>
  <div
    class="input default left"
  >
    <input
      type="number"
      value="2"
    />
  </div>
</div>
`;

exports[`Render with value of "undefined" 1`] = `
<div
  class="numberFieldFilterType"
>
  <div
    class="select"
    role="none"
  >
    <button
      class="displayValue default"
      type="button"
    >
      <div
        aria-label="sulu_admin.please_choose"
        class="croppedText"
        title="sulu_admin.please_choose"
      >
        <div
          aria-hidden="true"
          class="front"
        >
          sulu_admin.p
        </div>
        <div
          aria-hidden="true"
          class="back"
        >
          <span>
            lease_choose
          </span>
        </div>
        <div
          class="whole"
        >
          sulu_admin.please_choose
        </div>
      </div>
      <span
        aria-label="su-angle-down"
        class="su-angle-down toggle"
      />
    </button>
  </div>
  <div
    class="input default left"
  >
    <input
      type="number"
      value=""
    />
  </div>
</div>
`;
